### PR TITLE
Add OptionPicker

### PR DESCRIPTION
**Pull request by FilipMessa**
Added the PlacePicker, this is not pixel perfect solution still need some iteration, but this is PR is mainly focused on functionality. 

In dev mode may appear, issue with the performance but it should work OK in build(production version)

for next iterations
- improved designs 
- improved tests
- should be moved to margarita

screen shot 2019-01-30 at 10 11 42
screen shot 2019-01-30 at 10 10 16

### CODE REVIEW

**Comment by michalsanger** [comment]
Could selected items on web have this design (as seen on current kiwi.com homepage)?

screen shot 2019-01-30 at 10 30 14

**Comment by FilipMessa** [comment]
@michalsanger sure but first I would like to finish this basic stuff and then iterated on it.

**Comment by jvaclavik** [comment]
When you click on "Prešov" (its plus button), it adds to the input but remains in the list. Rest of the places (Prague, Berlin,...) works fine and it hides after the click.

File: src/TagsInput/TagsInput.js [hunk 1]
```diff
@@ -11,156 +16,159 @@ import InputField from './components/InputField';
...
+              onBlur={this.handleOnBlur}
+              ref={this.inputRef}
+              fontSize={fontSize}
+              disabled={disabled}
+              value={value}
+              placeholder={this.getPlaceholder()}
+              onChangeText={this.handleChange}
+            />
+          </ScrollView>
+          <DeleteButton
```
Comment (jvaclavik) [inline src/TagsInput/TagsInput.js]: I'm not sure if this button is OK from the UX perspective. It's quite unexpected that you can see this button only when you have something typed on keyboard and you can't delete with this button selected places from the list. 

I think it should works like that you can delete any place with this button (typed or selected) and it should be visible all the time except blank input. In this case also icon should be different, something like that: https://image.shutterstock.com/image-vector/delete-backspace-keyboard-key-line-450w-273565205.jpg

Current icon with cross would be also fine if we could delete all the content of the input with it. That is also approach which use Kiwi.com app. Maybe it would be the easiest change.

It's nothing serious for this PR, but we should keep it on mind in the future when we will improve UX and design of this component.

File: src/PlacePicker/__tests__/PlacePicker.test.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import { shallow, render, fireEvent } from 'react-native-testing-library';
+import PlacePicker, { filterSuggestedPlaces } from '../PlacePicker';
+import suggestedPlaces, { Prague, Berlin } from '../mocks/places';
+import { Badge } from '../../Badge';
+
+test('filterSuggestedPlaces', () => {
+  expect(filterSuggestedPlaces([Prague, Berlin], [Berlin])).toEqual([Prague]);
+});
+
+describe('#PlacePicker', () => {
```
Comment (jvaclavik) [inline src/PlacePicker/__tests__/PlacePicker.test.js]: What for is the `#`? Is it some convention?

File: src/TagsInput/TagsInput.js [hunk 3]
```diff
@@ -11,156 +16,159 @@ import InputField from './components/InputField';
...
+  return null;
 };
 
 export default class TagInput extends React.Component<Props, State> {
-  inputRef: ?{ current: null | React.Element<typeof TextInput> };
+  inputRef: any;
+  scrollRef: any;
+
   static defaultProps = {
     fontSize: 16,
```
Comment (jvaclavik) [inline src/TagsInput/TagsInput.js]: Maybe we could use design tokens here, probably `fontSizeButtonLarge`. It's some kind of button, right?

File: src/TagsInput/components/InputField.js [hunk 4]
```diff
@@ -4,71 +4,90 @@ import * as React from 'react';
...
-  componentWidth: number | 'auto',
+  componentWidth: number,
 |};
 
-class InputField extends React.PureComponent<Props, State> {
+class InputField extends React.Component<Props, State> {
   static defaultProps = {
-    maxWidth: 'auto',
-    minWidth: 'auto',
+    fontSize: 16,
```
Comment (jvaclavik) [inline src/TagsInput/components/InputField.js]: I think we could use here `fontSizeTextLarge` from the design tokens.

File: src/TagsInput/components/TagsContainer.js [hunk 5]
```diff
@@ -1,76 +1,59 @@
...
-  +maxWidth: 'auto' | ?number,
-  +minWidth: number,
 |};
 
-class TagsContainer extends React.PureComponent<Props> {
+export default class TagsContainer extends React.Component<Props> {
   static defaultProps = {
-    maxWidth: 'auto',
+    tags: [],
+    fontSize: 16,
```
Comment (jvaclavik) [inline src/TagsInput/components/TagsContainer.js]: I think we could use here `fontSizeTextLarge` from the design tokens.

**Comment by jvaclavik** [comment]
screenshot 2019-01-30 at 12 10 01

Positioning of "arrow" is not proper on Android

File: src/PlacePicker/PlacePicker.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  value: string,
+};
+
+type Event = { nativeEvent: { key: string } };
+
+export const filterSuggestedPlaces = (
+  suggestedPlaces: PlaceType[],
+  selectedPlaces: PlaceType[]
+) => {
+  const slectedIDs = selectedPlaces.map(place => place.id);
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: I am always making typos :see_no_evil:
Thanks

**Comment by FilipMessa** [comment]
> Positioning of "arrow" is not proper on Android

this is related to the RowOption component I guess, but good point I'll check and create the issue for that.

**Comment by FilipMessa** [comment]
> When you click on "Prešov" (its plus button), it adds to the input but remains in the list. Rest of the places (Prague, Berlin,...) works fine and it hides after the click.

should be fixed

File: src/TagsInput/components/InputField.js (hunk 4 above)
Comment (FilipMessa) [reply src/TagsInput/components/InputField.js]: done

File: src/TagsInput/TagsInput.js (hunk 3 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: done

File: src/PlacePicker/__tests__/PlacePicker.test.js (hunk 2 above)
Comment (FilipMessa) [reply src/PlacePicker/__tests__/PlacePicker.test.js]: done

File: src/TagsInput/TagsInput.js (hunk 1 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: done

File: src/PlacePicker/PlacePicker.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: done

**Comment by jvaclavik** [comment]
> > When you click on "Prešov" (its plus button), it adds to the input but remains in the list. Rest of the places (Prague, Berlin,...) works fine and it hides after the click.
> 
> should be fixed

Prešov is fine now, but I can multiple click on "sub places" like "Václav Havel Airport" or "Zoo...". Is it intented? I think it wasn't possible before.

screenshot 2019-01-30 at 13 01 04

File: src/TagsInput/TagsInput.js (hunk 1 above)
Comment (jvaclavik) [reply src/TagsInput/TagsInput.js]: Now it's much better, thanks!

**Comment by FilipMessa** [comment]
> Prešov is fine now, but I can multiple click on "sub places" like "Václav Havel Airport" or "Zoo...". Is it intented? I think it wasn't possible before.

fixed

File: src/PlacePicker/PlacePicker.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+import { StyleSheet } from '../PlatformStyleSheet';
+import { TagsInput } from '../TagsInput';
+import PlacePickerList from './components/PlacePickerList';
+
+type Props = {
+  onChangeText: (value: string) => void,
+  suggestedPlaces: PlaceType[],
+  onPressAdd?: (place: Array<PlaceType>) => void,
+  onPressItem?: (place: Array<PlaceType>) => void,
+  onRemoveSelectedPlace?: () => void,
```
Comment (JoseDuSV) [inline src/PlacePicker/PlacePicker.js]: Do we even need `onPressAdd `, `onPressItem` and `onRemoveSelectedPlace`?
I'm not 100% sure how will final implementation look like. But what about having just something like `onPlacesChange` which will be called on every add/remove/clear with selected places array in parameter.

File: package.json [hunk 8]
```diff
@@ -116,4 +116,4 @@
       }
     }
   }
-}
```
Comment (RobinCsl) [inline package.json]: Irrelevant change, can you please remove it from the PR?

File: src/PlacePicker/PlacePicker.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+      () => {
+        const { selected } = this.state;
+        onPressItem?.(selected);
+      }
+    );
+  };
+
+  handlePressItem = (place: PlaceType) => {
+    const { onPressAdd } = this.props;
+    this.setState({ selected: [place] }, () => {
```
Comment (RobinCsl) [inline src/PlacePicker/PlacePicker.js]: So here you are completely overriding the `selected` state with `place`, when in the previous method you just append `place`. Is that the expected behaviour?

File: src/PlacePicker/components/PlacePickerItem.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+function RenderDeparturePlaces({
+  places,
+  onAddPress,
+  onItemPress,
+}: {
+  places: Array<Place>,
+  onAddPress: (place: Place) => void,
+  onItemPress: (place: Place) => void,
+}) {
+  return places.map((p, index) => {
```
Comment (RobinCsl) [inline src/PlacePicker/components/PlacePickerItem.js]: Can `p` be renamed to `place`, please?

File: package.json (hunk 8 above)
Comment (FilipMessa) [reply package.json]: 👍

File: src/PlacePicker/PlacePicker.js (hunk 7 above)
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: I will rethink this but I am sure that we will need to have difference callback for `onPressAdd` and `onPressItem`. Because when the user clicks on the `onPressItem` he should be redirected to search with just one selected place.
And if the user clicks to `onPressAdd` it should update Search context and add the Tag to input.

File: src/PlacePicker/PlacePicker.js (hunk 9 above)
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: Yes, it is expected behavior, when user click on `onPressItem` it should select just one place and redirect to the search form.

If user clicks on the `onPressAdd` it should add the city to selection, as a new Tag, update search context but the user won't be redirected back.

File: src/PlacePicker/components/PlacePickerItem.js (hunk 10 above)
Comment (FilipMessa) [reply src/PlacePicker/components/PlacePickerItem.js]: 👍

File: src/PlacePicker/components/PlacePickerList.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+export default class PlacePickerList extends React.Component<Props> {
+  shouldComponentUpdate(nextProps: Props) {
+    const { places } = this.props;
+    const nextPlacesIds = getPlacesIds(nextProps.places);
+    const placesIds = getPlacesIds(places);
+
+    // compare the new-places IDs vs the old-places IDs and rerender if IDs changed.
+    return (
+      !placesIds.every(e => nextPlacesIds.includes(e)) ||
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/components/PlacePickerList.js]: Good point!

File: src/PlacePicker/PlacePicker.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+export const filterSuggestedPlaces = (
+  suggestedPlaces: PlaceType[],
+  selectedPlaces: PlaceType[]
+) => {
+  const selectedIDs = selectedPlaces.map(place => place.id);
+  return suggestedPlaces.reduce((acc = [], place) => {
+    if (selectedIDs.includes(place.id)) {
+      return acc;
+    }
+    if (place.departurePoints) {
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: Sure I will try to make it more readable. Thanks for comments

File: src/PlacePicker/PlacePicker.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+export default class PlacePicker extends React.Component<Props, State> {
+  state = {
+    selected: [],
+    value: '',
+  };
+
+  handlePressAdd = (place: PlaceType) => {
+    const { onPressItem } = this.props;
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: 🙈

File: package.json (hunk 8 above)
Comment (FilipMessa) [reply package.json]: done

File: src/PlacePicker/PlacePicker.js (hunk 7 above)
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: anyway I like the idea with `onPlacesChange` callback is more clear how to track changes, so I added callback jus for that. And simplify others.

File: src/PlacePicker/components/PlacePickerItem.js (hunk 10 above)
Comment (FilipMessa) [reply src/PlacePicker/components/PlacePickerItem.js]: done

File: src/PlacePicker/components/PlacePickerList.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/components/PlacePickerList.js]: done

File: src/PlacePicker/PlacePicker.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: Check it now pls, If it is better?

File: src/PlacePicker/PlacePicker.js (hunk 7 above)
Comment (JoseDuSV) [reply src/PlacePicker/PlacePicker.js]: Thanks for the explanation about `pressItem/addItem`. It makes sense to me now.
And thanks for adding `onChange`.

File: src/PlacePicker/components/PlacePickerItem.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        type={place.type}
+        header={place.name}
+        subheader={place.message}
+        onItemPress={createPressHandler(onItemPress, place)}
+        onAddPress={createPressHandler(onAddPress, place)}
+        border={borderType}
+      />
+      {departurePoints && (
+        <RenderDeparturePlaces
+          onItemPress={onItemPress}
```
Comment (RobinCsl) [inline src/PlacePicker/components/PlacePickerItem.js]: Will there not be a conflict between which onItemPress is going to be called?

File: src/PlacePicker/components/PlacePickerItem.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+import { RowOption } from '../../RowOption';
+import type { PlaceType } from '../PlacePickerTypes';
+
+type Props = {|
+  +place: PlaceType,
+  +onItemPress: (place: PlaceType) => void,
+  +onAddPress: (place: PlaceType) => void,
+|};
+
+function createPressHandler(func, place) {
```
Comment (RobinCsl) [inline src/PlacePicker/components/PlacePickerItem.js]: Could you move it to helpers maybe?

File: src/TagsInput/TagsInput.js [hunk 16]
```diff
@@ -11,156 +16,154 @@ import InputField from './components/InputField';
...
-  +selected?: string[],
+  +tags: string[],
   +disabled?: boolean,
-  +fontSize: number,
+  +label?: string,
   +onClearPress?: () => void,
+  +onKeyPress?: (e: Event) => void,
+  +placeholder?: string,
   +value?: string,
-  +autoFocus: boolean,
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: Why did you remove this prop?

File: src/TagsInput/TagsInput.js [hunk 17]
```diff
@@ -11,156 +16,154 @@ import InputField from './components/InputField';
...
+    return {
+      boxShadow: isFocus
+        ? `${defaultTokens.borderColorInputFocus} 0 0 0 2px inset`
+        : `${defaultTokens.borderColorInput} 0 0 0 1px inset`,
+    };
+  }
+  return null;
 };
 
 export default class TagInput extends React.Component<Props, State> {
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: The filename is Tag**s**Input, could you fix the typo in the name of the class?

File: src/TagsInput/TagsInput.js [hunk 18]
```diff
@@ -11,156 +16,154 @@ import InputField from './components/InputField';
...
-    Platform.OS === 'web' && this.inputRef.current.focus();
-
-    onClearPress?.();
-  };
-
   handleOnFocus = () => {
+    const { isFocus } = this.state;
+    if (isFocus) return;
     this.setState({ isFocus: true });
+    this.inputRef.current && this.inputRef.current.focus();
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]:
```suggestion
    this.inputRef.current?.focus.?();
```
or something like that?

Comment (RobinCsl) [reply src/TagsInput/TagsInput.js]: And maybe you could add it as callback after you set the state.

File: src/TagsInput/components/DeleteButton.js [hunk 19]
```diff
@@ -4,25 +4,24 @@ import * as React from 'react';
...
 import { defaultTokens } from '@kiwicom/orbit-design-tokens';
 import { Touchable } from '../../Touchable';
-import type { StylePropType } from '../../PlatformStyleSheet/StyleTypes';
 
 import { Icon } from '../../Icon';
 
 type Props = {|
   +onPress: () => void,
-  +style: StylePropType,
+  +opacity: number,
```
Comment (RobinCsl) [inline src/TagsInput/components/DeleteButton.js]: I would still leave the style as prop, it will be more extensible in the future.

File: src/TagsInput/components/InputField.js [hunk 20]
```diff
@@ -3,72 +3,92 @@
...
   }
 
-  handleOnLayout = (event: ViewLayoutEvent) => {
-    const { onLayout, maxWidth, minWidth, fontSize } = this.props;
-    const { width: inputWidth } = event.nativeEvent.layout;
+  shouldComponentUpdate(nextProps, nextState) {
+    const { fontSize, maxWidth, minWidth, placeholder, value } = this.props;
+
+    if (
+      fontSize !== nextProps.fontSize ||
```
Comment (RobinCsl) [inline src/TagsInput/components/InputField.js]: Could you give names or group the conditions to make it more understandable why these conditions are needed?

File: src/TagsInput/components/InputField.js [hunk 21]
```diff
@@ -3,72 +3,92 @@
...
+      maxWidth !== nextProps.maxWidth ||
+      placeholder !== nextProps.placeholder ||
+      minWidth ||
```
Comment (RobinCsl) [inline src/TagsInput/components/InputField.js]: Should this be `minWidth !== nextProps.minWidth`?

Comment (RobinCsl) [reply src/TagsInput/components/InputField.js]: I believe we have the fast-compare library, so you could probably just do `isEqual(props, nextProps)`?

File: src/TagsInput/components/InputField.js [hunk 22]
```diff
@@ -77,47 +97,49 @@ class InputField extends React.PureComponent<Props, State> {
...
+
+    const dynamicStyle = {
+      common: { fontSize },
       input: {
         width: componentWidth,
       },
-    });
+    };
+
+    const sizerValue = value || (placeholder || '');
```
Comment (RobinCsl) [inline src/TagsInput/components/InputField.js]: What do you mean by `sizer`?

**Comment by RobinCsl** [comment]
Quite a big PR! I added a few comments in the code.
One thing that strikes me now is that it seems to be _too specific_ for `universal-components`. I would change the naming to `OptionPicker` or something similar. Also, the shape of your data should be more agnostic than `departurePoint` or things like this. 
That means the final implementation in `margarita` will use most of this PR, but this component should do a bit less and just provide a way to pick different options, with their sub-options, and not necessarily places as such.
So a bit of refactoring in perspective in my opinion.

**Comment by FilipMessa** [comment]
@RobinCsl 	
I know that is big and I am sorry about that.

> One thing that strikes me now is that it seems to be too specific for universal-components

I had never in my plan keep this component in the universal-components, my plan was moved it into the margarita later.
I developed it here because I expected that I need to make some changes in the TagsInput.

I'll consider your suggestion, but I am not sure about that now.

BTW: thanks for your CR, I will jump on your note.

File: src/TagsInput/TagsInput.js (hunk 18 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: I will consider the callback.

About the `optional-chaining`, I tried it before, but flow starts to scream on me this:
`Flow does not yet support method or property calls in optional chains.`

File: src/TagsInput/components/InputField.js (hunk 21 above)
Comment (FilipMessa) [reply src/TagsInput/components/InputField.js]: Thanks, I will look at this library.a

File: src/TagsInput/components/InputField.js (hunk 22 above)
Comment (FilipMessa) [reply src/TagsInput/components/InputField.js]: In my thought, I call like that, this Text helper component (line 128), which is used for width computation for TextInput.

File: src/PlacePicker/PlacePicker.js [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+    return (
+      <View>
+        <View style={styles.inputWrapper}>
+          <TagsInput
+            tags={tags}
+            onClearPress={this.handleClear}
+            value={value}
+            onChangeText={this.handleChangeText}
+            onKeyPress={this.handleBackspacePress}
+            label="From:"
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/PlacePicker/PlacePicker.js]: I don't understand what do you mean with this note.

File: src/PlacePicker/components/PlacePickerItem.js (hunk 15 above)
Comment (FilipMessa) [reply src/PlacePicker/components/PlacePickerItem.js]: This is the helper just for the PlacePickerItem so I don't think we would like to use it anywhere else. So I don't see the reason why put this to helpers if it is and will be used it just inside PlacePicker.

File: src/TagsInput/TagsInput.js (hunk 16 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: I will add this props back I don't know.

File: src/TagsInput/TagsInput.js (hunk 17 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: 👍 thanks

File: src/TagsInput/components/InputField.js (hunk 21 above)
Comment (FilipMessa) [reply src/TagsInput/components/InputField.js]: I can't find in the `package.js` this dependency.

Comment (tbergq) [reply src/TagsInput/components/InputField.js]: I think it is in packages/utils

Comment (tbergq) [reply src/TagsInput/components/InputField.js]: Of course that is in margarita

**Comment by FilipMessa** [comment]
@RobinCsl pls check again thanks

File: src/OptionPicker/OptionPicker.js [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+    }
+  };
+
+  render() {
+    const { selected, value } = this.state;
+    const { options, label, placeholder } = this.props;
+    const tags = selected.map(option => option.label);
+
+    return (
+      <View>
```
Comment (RobinCsl) [inline src/OptionPicker/OptionPicker.js]: This might need to be a ScrollView if the list is pretty long.

Comment (FilipMessa) [reply src/OptionPicker/OptionPicker.js]: 👍 good point

Comment (FilipMessa) [reply src/OptionPicker/OptionPicker.js]: Maybe will be better add the ScrollView just for Option List, so input filed will stay on place.

File: src/TagsInput/TagsInput.js [hunk 25]
```diff
@@ -11,156 +16,163 @@ import InputField from './components/InputField';
...
+  return null;
+};
+
+export default class TagsInput extends React.Component<Props, State> {
+  inputRef: any;
+  scrollRef: any;
+
   static defaultProps = {
-    fontSize: 16,
-    autoFocus: true,
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: Could you put this back in?

File: src/TagsInput/TagsInput.js [hunk 26]
```diff
@@ -11,156 +16,163 @@ import InputField from './components/InputField';
...
   constructor(props: Props) {
     super(props);
     this.inputRef = React.createRef();
+    this.scrollRef = React.createRef();
 
     this.state = {
-      value: props.value ?? '',
-      containerWidth: null,
-      inputWidth: null,
-      isFocus: props.autoFocus,
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: Likewise here.

File: src/TagsInput/TagsInput.js [hunk 27]
```diff
@@ -11,156 +16,163 @@ import InputField from './components/InputField';
...
-            autoFocus={isFocus}
+      <TouchableWithoutFeedback onPress={this.handleOnFocus}>
+        <View style={[styles.container, dynamicStyle.border]}>
+          {Boolean(label) && (
+            <Text weight="bold" style={[styles.label, dynamicStyle.label]}>
+              {label}
+            </Text>
+          )}
+          <ScrollView
+            contentContainerStyle={{ flexGrow: 1 }}
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: This style could be extracted.

File: src/TagsInput/TagsInput.stories.js [hunk 28]
```diff
@@ -18,18 +18,16 @@ storiesOf('TagsInput', module)
         onChangeText={action('onChangeText')}
         label="From:"
         placeholder="Departure point"
-        autoFocus={false}
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.stories.js]: Could you add this back in?

File: src/TagsInput/TagsInput.stories.js [hunk 29]
```diff
@@ -18,18 +18,16 @@ storiesOf('TagsInput', module)
...
   ))
   .add('with Tags', () => (
     <View style={{ margin: 20 }}>
       <TagsInput
-        selected={selectedMock}
+        tags={selectedMock}
         onChangeText={action('onChangeText')}
         label="From:"
         placeholder="Departure point"
-        autoFocus={false}
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.stories.js]: Same here.

File: src/TagsInput/components/InputField.js [hunk 30]
```diff
@@ -77,47 +98,50 @@ class InputField extends React.PureComponent<Props, State> {
...
           ref={forwardRef}
           editable={!disabled}
-          autoFocus={autoFocus}
           onChangeText={onChangeText}
           value={value}
           placeholder={placeholder}
-          style={[styles.input, dynamicStyle.common, dynamicStyle.input]}
+          style={[styles.input, dynamicStyle.input, dynamicStyle.common]}
         />
+        <Text
```
Comment (RobinCsl) [inline src/TagsInput/components/InputField.js]: So we're displaying something that's opacity 0? I am not sure this is going to be great for accessibility.

File: src/TagsInput/TagsInput.js (hunk 26 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: `isFocus: props.autofocus || false,` I will change it like this so we don't need a set default value. Or you still prefer to have it?

File: src/TagsInput/TagsInput.js [hunk 31]
```diff
@@ -11,156 +16,163 @@ import InputField from './components/InputField';
...
+              ref={this.inputRef}
+              fontSize={fontSize}
+              disabled={disabled}
+              value={value}
+              placeholder={this.getPlaceholder()}
+              onChangeText={this.handleChange}
+            />
+          </ScrollView>
+          <DeleteButton
+            opacity={dynamicStyle.deleteButton.opacity}
```
Comment (RobinCsl) [inline src/TagsInput/TagsInput.js]: This looks rather complex. I believe `style` would have been easier, because you could write
`style={dynamicStyle.deleteButton}`.
Besides, your styles don't need to be fully dynamic:
`style={[style.opacityZero, tags.length > 0 && style.opacityOne, value && dynamicStyle.deleteButton]}`

File: src/TagsInput/components/DeleteButton.js (hunk 19 above)
Comment (RobinCsl) [reply src/TagsInput/components/DeleteButton.js]: I am still not convinced by that. It is not consistent with the rest of the codebase. If I want to reuse that component, I should be able to. I don't see why we should prevent re-use.

File: src/TagsInput/components/TagsContainer.js [hunk 32]
```diff
@@ -1,76 +1,60 @@
...
-          </View>
-        </ScrollView>
-      </View>
+      <>
+        {tags.map(tag => (
+          <Badge
+            fontSize={fontSize}
+            type="info"
+            key={tag}
+            testID="input-tag"
```
Comment (RobinCsl) [inline src/TagsInput/components/TagsContainer.js]: All the tags have the same `testID`. Is that expected?

File: src/TagsInput/TagsInput.js [hunk 33]
```diff
@@ -11,156 +16,163 @@ import InputField from './components/InputField';
...
-            minWidth={INPUT_MIN_WIDTH}
-            maxWidth={this.subtractFromContainerWidth(TAGS_MIN_WIDTH)}
-            onLayout={this.setInputWidth}
-            value={value}
-            placeholder={this.getPlaceholder()}
-            onChangeText={this.handleChange}
-            autoFocus={isFocus}
+      <TouchableWithoutFeedback onPress={this.handleOnFocus}>
+        <View style={[styles.container, dynamicStyle.border]}>
+          {Boolean(label) && (
```
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: Sure but if the label value will be the empty string you will render unnecessary empty element.

File: src/TagsInput/TagsInput.stories.js (hunk 28 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.stories.js]: 👍 thanks

File: src/TagsInput/TagsInput.stories.js (hunk 29 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.stories.js]: Sure

File: src/TagsInput/components/TagsContainer.js (hunk 32 above)
Comment (FilipMessa) [reply src/TagsInput/components/TagsContainer.js]: Good point, I didn't realize before that this is not the best practice, but I already read about that, at documentation react-test-library, that every render Id should be unique. I will change that

File: src/TagsInput/components/InputField.js (hunk 30 above)
Comment (FilipMessa) [reply src/TagsInput/components/InputField.js]: Me either, but there is nothing like `visibility: 'hidden',` for react-native
but I will add there `accessibilityRole="none"` this should probably fix this issue.

File: src/TagsInput/TagsInput.js (hunk 31 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: > `style={dynamicStyle.deleteButton}`.
Besides, your styles don't need to be fully dynamic:
`style={[style.opacityZero, tags.length > 0 && style.opacityOne, value && dynamicStyle.deleteButton]}`

💯

File: src/TagsInput/TagsInput.js (hunk 33 above)
[the remark this answers is not in the document]
Comment (RobinCsl) [reply src/TagsInput/TagsInput.js]: Okay then. But maybe I would be really verbose and write `(label != null || label !== '') && (...)`, the condition is then clearer than using Boolean.

File: src/TagsInput/TagsInput.js (hunk 26 above)
Comment (RobinCsl) [reply src/TagsInput/TagsInput.js]: Well, we have `defaultProps` anyway, so I would put it there. Otherwise, we have default values spread in different places.

File: src/TagsInput/TagsInput.js (hunk 33 above)
[the remark this answers is not in the document]
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: I prefer `!!label && ...` but I'll change it as you suggested.

File: src/TagsInput/TagsInput.js (hunk 26 above)
Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: OK

Comment (FilipMessa) [reply src/TagsInput/TagsInput.js]: make sense
